File: python-client-sdk/src/orbapi/models/operations/get_customer_costs_by_id.py
import dataclasses
from datetime import date, datetime
from marshmallow import fields
import dateutil.parser
from typing import Any,Optional
from enum import Enum
from dataclasses_json import dataclass_json
from orbapi import utils
from ..shared import price as shared_price


@dataclasses.dataclass
class GetCustomerCostsByIDPathParams:
    customer_id: str = dataclasses.field(metadata={'path_param': { 'field_name': 'customer_id', 'style': 'simple', 'explode': False }})
    
class GetCustomerCostsByIDViewModeEnum(str, Enum):
    PERIODIC = "periodic"
    CUMULATIVE = "cumulative"


@dataclasses.dataclass
class GetCustomerCostsByIDQueryParams:
    group_by: Optional[str] = dataclasses.field(default=None, metadata={'query_param': { 'field_name': 'group_by', 'style': 'form', 'explode': True }})
    timeframe_end: Optional[str] = dataclasses.field(default=None, metadata={'query_param': { 'field_name': 'timeframe_end', 'style': 'form', 'explode': True }})
    timeframe_start: Optional[datetime] = dataclasses.field(default=None, metadata={'query_param': { 'field_name': 'timeframe_start', 'style': 'form', 'explode': True }})
    view_mode: Optional[GetCustomerCostsByIDViewModeEnum] = dataclasses.field(default=None, metadata={'query_param': { 'field_name': 'view_mode', 'style': 'form', 'explode': True }})
    

@dataclass_json
@dataclasses.dataclass
class GetCustomerCostsByID200ApplicationJSONDataPerPriceCostsPriceGroups:
    grouping_key: str = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('grouping_key') }})
    grouping_value: str = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('grouping_value') }})
    total: str = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('total') }})
    secondary_grouping_key: Optional[str] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('secondary_grouping_key') }})
    secondary_grouping_value: Optional[str] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('secondary_grouping_value') }})
    

@dataclass_json
@dataclasses.dataclass
class GetCustomerCostsByID200ApplicationJSONDataPerPriceCosts:
    r"""GetCustomerCostsByID200ApplicationJSONDataPerPriceCosts
    Price's contributions for the timeframe, excluding any transforms (minimums and discounts).
    """
    
    price_groups: list[GetCustomerCostsByID200ApplicationJSONDataPerPriceCostsPriceGroups] = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('price_groups') }})
    price: Optional[shared_price.Price] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('price') }})
    subtotal: Optional[str] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('subtotal') }})
    total: Optional[str] = dataclasses.field(default=None, metadata={'dataclasses_json': { 'letter_case': utils.field_name('total') }})
    

@dataclass_json
@dataclasses.dataclass
class GetCustomerCostsByID200ApplicationJSONData:
    per_price_costs: list[GetCustomerCostsByID200ApplicationJSONDataPerPriceCosts] = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('per_price_costs') }})
    subtotal: str = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('subtotal') }})
    timeframe_end: datetime = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('timeframe_end'), 'encoder': utils.datetimeisoformat(False), 'decoder': dateutil.parser.isoparse, 'mm_field': fields.DateTime(format='iso') }})
    timeframe_start: datetime = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('timeframe_start'), 'encoder': utils.datetimeisoformat(False), 'decoder': dateutil.parser.isoparse, 'mm_field': fields.DateTime(format='iso') }})
    total: str = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('total') }})
    

@dataclass_json
@dataclasses.dataclass
class GetCustomerCostsByID200ApplicationJSON:
    data: list[GetCustomerCostsByID200ApplicationJSONData] = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('data') }})
    pagination_metadata: dict[str, Any] = dataclasses.field(metadata={'dataclasses_json': { 'letter_case': utils.field_name('pagination_metadata') }})
    

@dataclasses.dataclass
class GetCustomerCostsByIDRequest:
    path_params: GetCustomerCostsByIDPathParams = dataclasses.field()
    query_params: GetCustomerCostsByIDQueryParams = dataclasses.field()
    

@dataclasses.dataclass
class GetCustomerCostsByIDResponse:
    content_type: str = dataclasses.field()
    status_code: int = dataclasses.field()
    get_customer_costs_by_id_200_application_json_object: Optional[GetCustomerCostsByID200ApplicationJSON] = dataclasses.field(default=None)
    

File: typescript-client-sdk/src/sdk/models/operations/getcustomercostsbyid.ts
import { SpeakeasyBase, SpeakeasyMetadata } from "../../../internal/utils";
import * as shared from "../shared";
import { Expose, Transform, Type } from "class-transformer";


export class GetCustomerCostsByIdPathParams extends SpeakeasyBase {
  @SpeakeasyMetadata({ data: "pathParam, style=simple;explode=false;name=customer_id" })
  customerId: string;
}
export enum GetCustomerCostsByIdViewModeEnum {
    Periodic = "periodic",
    Cumulative = "cumulative"
}

export class GetCustomerCostsByIdQueryParams extends SpeakeasyBase {
  @SpeakeasyMetadata({ data: "queryParam, style=form;explode=true;name=group_by" })
  groupBy?: string;

  @SpeakeasyMetadata({ data: "queryParam, style=form;explode=true;name=timeframe_end" })
  timeframeEnd?: string;

  @SpeakeasyMetadata({ data: "queryParam, style=form;explode=true;name=timeframe_start;dateTimeFormat=YYYY-MM-DDThh:mm:ss.sssZ" })
  timeframeStart?: Date;

  @SpeakeasyMetadata({ data: "queryParam, style=form;explode=true;name=view_mode" })
  viewMode?: GetCustomerCostsByIdViewModeEnum;
}

export class GetCustomerCostsByIdRequest extends SpeakeasyBase {
  @SpeakeasyMetadata()
  pathParams: GetCustomerCostsByIdPathParams;

  @SpeakeasyMetadata()
  queryParams: GetCustomerCostsByIdQueryParams;
}

export class GetCustomerCostsById200ApplicationJSONDataPerPriceCostsPriceGroups extends SpeakeasyBase {
  @SpeakeasyMetadata()
  @Expose({ name: "grouping_key" })
  groupingKey: string;

  @SpeakeasyMetadata()
  @Expose({ name: "grouping_value" })
  groupingValue: string;

  @SpeakeasyMetadata()
  @Expose({ name: "secondary_grouping_key" })
  secondaryGroupingKey?: string;

  @SpeakeasyMetadata()
  @Expose({ name: "secondary_grouping_value" })
  secondaryGroupingValue?: string;

  @SpeakeasyMetadata()
  @Expose({ name: "total" })
  total: string;
}

// GetCustomerCostsById200ApplicationJSONDataPerPriceCosts
/** 
 * Price's contributions for the timeframe, excluding any transforms (minimums and discounts).
**/
export class GetCustomerCostsById200ApplicationJSONDataPerPriceCosts extends SpeakeasyBase {
  @SpeakeasyMetadata()
  @Expose({ name: "price" })
  @Type(() => shared.Price)
  price?: shared.Price;

  @SpeakeasyMetadata({ elemType: GetCustomerCostsById200ApplicationJSONDataPerPriceCostsPriceGroups })
  @Expose({ name: "price_groups" })
  @Type(() => GetCustomerCostsById200ApplicationJSONDataPerPriceCostsPriceGroups)
  priceGroups: GetCustomerCostsById200ApplicationJSONDataPerPriceCostsPriceGroups[];

  @SpeakeasyMetadata()
  @Expose({ name: "subtotal" })
  subtotal?: string;

  @SpeakeasyMetadata()
  @Expose({ name: "total" })
  total?: string;
}

export class GetCustomerCostsById200ApplicationJSONData extends SpeakeasyBase {
  @SpeakeasyMetadata({ elemType: GetCustomerCostsById200ApplicationJSONDataPerPriceCosts })
  @Expose({ name: "per_price_costs" })
  @Type(() => GetCustomerCostsById200ApplicationJSONDataPerPriceCosts)
  perPriceCosts: GetCustomerCostsById200ApplicationJSONDataPerPriceCosts[];

  @SpeakeasyMetadata()
  @Expose({ name: "subtotal" })
  subtotal: string;

  @SpeakeasyMetadata()
  @Expose({ name: "timeframe_end" })
  @Transform(({ value }) => new Date(value), { toClassOnly: true })
  timeframeEnd: Date;

  @SpeakeasyMetadata()
  @Expose({ name: "timeframe_start" })
  @Transform(({ value }) => new Date(value), { toClassOnly: true })
  timeframeStart: Date;

  @SpeakeasyMetadata()
  @Expose({ name: "total" })
  total: string;
}

export class GetCustomerCostsById200ApplicationJSON extends SpeakeasyBase {
  @SpeakeasyMetadata({ elemType: GetCustomerCostsById200ApplicationJSONData })
  @Expose({ name: "data" })
  @Type(() => GetCustomerCostsById200ApplicationJSONData)
  data: GetCustomerCostsById200ApplicationJSONData[];

  @SpeakeasyMetadata()
  @Expose({ name: "pagination_metadata" })
  paginationMetadata: Record<string, any>;
}

export class GetCustomerCostsByIdResponse extends SpeakeasyBase {
  @SpeakeasyMetadata()
  contentType: string;

  @SpeakeasyMetadata()
  statusCode: number;

  @SpeakeasyMetadata()
  getCustomerCostsById200ApplicationJSONObject?: GetCustomerCostsById200ApplicationJSON;
}